flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java

```diff
@@ -0,0 +1,432 @@
+
+    @RegisterExtension
+    private final EachCallbackWrapper<ZooKeeperExtension> eachWrapper =
+            new EachCallbackWrapper<>(zooKeeperExtension);
+
+    @Test
+    public void testElectionDriverGainsLeadership() throws Exception {
+        final TestingLeaderElectionListener leaderElectionListener =
+                new TestingLeaderElectionListener();
+        final CuratorFrameworkWithUnhandledErrorListener curatorFramework = startCuratorFramework();
+
+        final ZooKeeperMultipleComponentLeaderElectionDriver leaderElectionDriver =
+                startLeaderElectionDriver(
+                        leaderElectionListener, curatorFramework.asCuratorFramework());
+
+        try {
+            leaderElectionListener.await(IsLeaderEvent.class);
+        } finally {
+            leaderElectionDriver.close();
+            curatorFramework.close();
+        }
+    }
+
+    @Test
+    public void testElectionDriverLosesLeadership() throws Exception {
+        final TestingLeaderElectionListener leaderElectionListener =
+                new TestingLeaderElectionListener();
+        final CuratorFrameworkWithUnhandledErrorListener curatorFramework = startCuratorFramework();
+
+        final ZooKeeperMultipleComponentLeaderElectionDriver leaderElectionDriver =
+                startLeaderElectionDriver(
+                        leaderElectionListener, curatorFramework.asCuratorFramework());
+
+        try {
+            leaderElectionListener.await(IsLeaderEvent.class);
+
+            zooKeeperExtension.stop();
+
+            leaderElectionListener.await(NotLeaderEvent.class);
+        } finally {
+            leaderElectionDriver.close();
+            curatorFramework.close();
+        }
+    }
+
+    @Test
+    public void testPublishLeaderInformation() throws Exception {
+        final TestingLeaderElectionListener leaderElectionListener =
+                new TestingLeaderElectionListener();
+        final CuratorFrameworkWithUnhandledErrorListener curatorFramework = startCuratorFramework();
+
+        final ZooKeeperMultipleComponentLeaderElectionDriver leaderElectionDriver =
+                startLeaderElectionDriver(
+                        leaderElectionListener, curatorFramework.asCuratorFramework());
+
+        try {
+            leaderElectionListener.await(IsLeaderEvent.class);
+
+            final String componentId = "barfoo";
+            final DefaultLeaderRetrievalService defaultLeaderRetrievalService =
+                    new DefaultLeaderRetrievalService(
+                            new ZooKeeperLeaderRetrievalDriverFactory(
+                                    curatorFramework.asCuratorFramework(),
+                                    componentId,
+                                    ZooKeeperLeaderRetrievalDriver.LeaderInformationClearancePolicy
+                                            .ON_LOST_CONNECTION));
+
+            final TestingListener leaderRetrievalListener = new TestingListener();
```

Comment:
True. Since this is unrelated to my changes I will leave this for a future follow up. I fear that making this renamings will screw up the fixing of other changes.